### PR TITLE
Prevent unbinding reused DOM elements while rendering.

### LINKED ISSUE #922: Explosion when playing with the basic styles
apr-18-2017 18-25-46

### PR DESCRIPTION
**Pull request by szymonkups**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Prevent unbinding elements that are reused during rendering. Closes #922.

BREAKING CHANGE: Removed `Renderer#getCorrespondingDom` method.
BREAKING CHANGE: Renamed `Renderer#getCorrespondingDomText` method to  `Renderer#findCorrespondingDomText`.
BREAKING CHANGE: Merged `Renderer#getCorrespondingDomElement` and `Renderer#getCorrespondingDomDocumentFragment` into one method `Renderer#mapViewToDom`

---

### Additional information

Close alongside with https://github.com/ckeditor/ckeditor5-link/tree/t/ckeditor5-engine/922.

### CODE REVIEW

**Comment by szymonkups** [comment]
I've only chaged view -> DOM methods. We could probably change DOM -> view methods as well:
 * remove `getCorrespondingView`
 * merge `getCorrespondingViewElement` and `getCorrespondingViewDocumentFragment` into `mapDomToView`
  * rename `getCorrespondingViewText` to `findCorrespondingViewText`.

If yes - can we move it to separate PR?

**Comment by Reinmar** [comment]
Still reproducible:

* https://github.com/ckeditor/ckeditor5-engine/issues/711#issuecomment-309722442
* https://github.com/ckeditor/ckeditor5/issues/453


Fixed:

* https://github.com/ckeditor/ckeditor5-basic-styles/issues/46

So it seems that the problem with selection is a separate problem. It's not crashing the editor and most likely it's only a rendering issue (see what happens when you test https://github.com/ckeditor/ckeditor5-basic-styles/issues/46 – selection is not rendered on the whole word but you apply a style to the whole word pressing <kbd>Ctrl</kbd>+<kbd>B</kbd>).

File: src/view/domconverter.js [hunk 1]
```diff
@@ -322,7 +322,10 @@ export default class DomConverter {
 
 				domAfter = domParent.childNodes[ 0 ];
 			} else {
-				domBefore = this.getCorrespondingDom( viewPosition.nodeBefore );
+				const nodeBefore = viewPosition.nodeBefore;
+				domBefore = nodeBefore.is( 'text' ) ?
```
Comment (pjasiun) [inline src/view/domconverter.js]: New line before is missing.

File: src/view/renderer.js [hunk 2]
```diff
@@ -464,20 +462,28 @@ export default class Renderer {
...
-				this.domConverter.unbindDomElement( actualDomChildren[ i ] );
+				nodesToUnbind.add( actualDomChildren[ i ] );
 				remove( actualDomChildren[ i ] );
 			} else { // 'equal'
 				i++;
 			}
 		}
 
+		// Unbind removed nodes. We don't need to check child nodes, because if child node was reinserted, it was moved
+		// to DOM tree out of the removed node.
```
Comment (pjasiun) [inline src/view/renderer.js]: I would add also comment why we need to check parent. It will not be clear even for us, in 3 mounts.

**Comment by Reinmar** [comment]
> If yes - can we move it to separate PR?

I think that since we changed some method names already in this PR it would be good to change the rest in this PR as well.

**Comment by szymonkups** [comment]
Reported a follow up: https://github.com/ckeditor/ckeditor5-engine/issues/991

**Comment by Reinmar** [comment]
Dooh... I realised just now that the API changes were in `DomConverter` not `Renderer` :D. Seconds after merging it.

**Pull request merged**